OpenSmc.Ifrs17/OpenSmc.Ifrs17.ImportHub/ParsingStorage.cs

```diff
@@ -0,0 +1,250 @@
+                    .Where(et => et.InputSource == InputSource.Opening) //This Contains overload cannot be used in DB, thus the ToArrayAsync()
+                    .Select(et => et.SystemName).ToHashSet(),
+                _ => Enumerable.Empty<string>().ToHashSet(),
+            };
+
+
+            // DimensionsWithExternalId
+            DimensionsWithExternalId = new Dictionary<Type, Dictionary<string, string>>()
+            {
+                { typeof(AmountType), GetDimensionWithExternalIdDictionary<AmountType>() },
+                { typeof(EstimateType), GetDimensionWithExternalIdDictionary<EstimateType>() }
+            };
+
+            HierarchicalDimensionCache.Initialize<AmountType>();
+        }
+
+        public Dictionary<string, string> GetDimensionWithExternalIdDictionary<T>()
+            where T : KeyedOrderedDimension
+        {
+            var dict = new Dictionary<string, string>();
+            var items = Workspace.GetData<T>().ToArray();
+            foreach (var item in items)
+            {
+                dict.TryAdd(item.SystemName, item.SystemName);
+                if (typeof(T).IsAssignableTo(typeof(KeyedOrderedDimensionWithExternalId)))
+                {
+                    var externalIds = (string[])(typeof(T)
+                        .GetProperty(nameof(KeyedOrderedDimensionWithExternalId.ExternalId)).GetValue(item));
+                    if (externalIds == null) continue;
+                    foreach (var extId in externalIds)
+                        dict.TryAdd(extId, item.SystemName);
+                }
+            }
+
+            return dict;
+        }
+
+        // Getters
+        public bool IsDataNodeReinsurance(string goc) => DataNodeDataBySystemName[goc].IsReinsurance;
+        public bool IsValidDataNode(string goc) => DataNodeDataBySystemName.ContainsKey(goc);
+
+        public CashFlowPeriodicity GetCashFlowPeriodicity(string goc)
+        {
+            if (!SingleDataNodeParametersByGoc.TryGetValue(goc, out var inner))
+                return CashFlowPeriodicity.Monthly;
+            return inner[Consts.CurrentPeriod].CashFlowPeriodicity;
+        }
+
+        public InterpolationMethod GetInterpolationMethod(string goc)
+        {
+            if (!SingleDataNodeParametersByGoc.TryGetValue(goc, out var inner))
+                return InterpolationMethod.NotApplicable;
+            return inner[Consts.CurrentPeriod].InterpolationMethod;
+        }
+
+        /*
+        // Validations
+        public string ValidateEstimateType(string et, string goc)
+        {
+            var allowedEstimateTypes = estimateTypes;
+            if (DataNodeDataBySystemName.TryGetValue(goc, out var dataNodeData) &&
+                dataNodeData.LiabilityType == LiabilityTypes.LIC)
+                estimateTypes.ExceptWith(TechnicalMarginEstimateTypes);
+            if (!allowedEstimateTypes.Contains(et))
+                ApplicationMessage.Log(Error.EstimateTypeNotFound, et);
```

Comment:
@amuolo Should Exceptions be thrown instead of logging errors?
I expect there is a try catch somewhere...